File: sem4/mpp/ex2/server/src/main/java/server/LoginService.java
package server;

import domain.User;
import repository.UserRepository;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import static java.security.MessageDigest.*;

public class LoginService {
	private final UserRepository userRepository;
	public final MessageDigest md;

	{
		try {
			md = getInstance("SHA-512");
		} catch (NoSuchAlgorithmException e) {
			throw new RuntimeException(e);
		}
	}

	private String hash(String salt, String text) {
		String saltedText = salt + text;
		byte[] bytes = md.digest(saltedText.getBytes(StandardCharsets.UTF_8));
		return new String(bytes, StandardCharsets.UTF_8);
	}

	public LoginService(UserRepository userDBRepository) {
		this.userRepository = userDBRepository;
	}

	public synchronized void register(String username, String password) {
		if (userRepository.getUserByUsername(username) != null)
			throw new RuntimeException("User already registered!");
		password = hash("piezisa", password);
		userRepository.add(new User(username, password));
	}

	public synchronized User login(String username, String password) {
		User user = userRepository.getUserByUsername(username);
		if (user == null)
			throw new RuntimeException("No such user exists!");
		password = hash("piezisa", password);
		if (!user.getHashedPassword().equals(password))
			throw new RuntimeException("Wrong password");
		return user;
	}
}

File: sem4/mpp/project-cs/server/service/LoginService.cs
using project_cs.domain;
using project_cs.repo;

namespace project_cs.service;

public class LoginService {
	private readonly IUserRepository _userRepository;

	public LoginService(IUserRepository userRepository) {
		_userRepository = userRepository;
	}

	private string Hash(string salt, string text) {
		var saltedText = salt + text;
		var bytes = System.Text.Encoding.UTF8.GetBytes(saltedText);
		using (var hash = System.Security.Cryptography.SHA512.Create()) {
			var hashedInputBytes = hash.ComputeHash(bytes);
			return System.Text.Encoding.UTF8.GetString(hashedInputBytes);
		}
	}

	public void Register(string username, string password) {
		if (_userRepository.GetUserByUsername(username) != null)
			throw new Exception("User already registered!");
		password = Hash("piezisa", password);
		_userRepository.Add(new User(username, password));
	}

	public User Login(string username, string password) {
		var user = _userRepository.GetUserByUsername(username);
		if (user == null)
			throw new Exception("No such user exists!");
		password = Hash("piezisa", password);
		if (user.hashedPassword != password)
			throw new Exception("Wrong password");
		return user;
	}
}